Documentation/technote-ocsp-cert-revocation.md

```diff
@@ -0,0 +1,290 @@
+are found to be compromised.
+
+**Other mechanisms**
+
+Another mechanism tat can be used is to issue certificates with short lifetimes. Connections are
+restricted to short periods without incurring the revocation cost / complexity.
+
+# Java, TLS and Certificate Revocation, Debug Options
+
+Before we look at demonstrations of OCSP and OCSP Stapling with Java and Daml, here are some
+techniques to debug the TLS connections and see what is happening on the wire.
+
+## To enable certificate revocation
+
+See Oracle documentation (link in **References**) around these options
+
+```
+# Add ocsp.enable=true into security properties. The following overrides
+-Djava.security.properties=$ROOTDIR/java.security
+
+# The following enables OCSP
+-Dcom.sun.net.ssl.checkRevocation=true 
+
+# Enable CRL checking
+-Dcom.sun.security.enableCRLDP=true
+
+# The following additionally enable OCSP Stapling
+-Djdk.tls.client.enableStatusRequestExtension=true 
+-Djdk.tls.server.enableStatusRequestExtension=true 
+```
+
+Implementation Notes:
+- Java JDK expects certificates to be in truststore format so TLS Client/Server example imports
+  into Truststores and Keystores. See ```run-tls-server.sh```
+
+## Java TLS Debug options
+
+Debugging TLS issues can be challenging. The following enables debug logging to the JDK
+TLS stack, particularly ```certpath``` (Certificate validation), ```ocsp``` (additional option for OCSP in certpath),
+```ssl:handshake``` (TLS process handling) and ```respmgr``` (debug Response Manager which
+handles OCSP responses and other items)
+
+```aidl
+-Djava.security.debug="certpath ocsp" 
+-Djavax.net.debug="ssl:handshake,verbose,respmgr" 
+```
+
+Implementation Notes:
+- JDK OCSP Checker expects the TLS certificate to contain the full CA trust chain with the
+  server cert first followed by intermediate and root. Just returning the server certificate without chain
+  will produce a "OCSP is Disabled" message in debug trace of respmgr. Concatenate the server and CS certs into the
+  certificate PEM file resolves the issue.
+
+## Wireshark and traffic captures
+
+To be able to see the traffic flows between client and server, it can help to use Wireshark. However TLS is encrypted by default
+and traffic data is hidden. This can be resolved by enabling TLS Session key logging. See **References** for links
+to details articles and setup.
+
+For the JDK, you can use the jSSLKeyLog agent to capture the TLS session key to a file and then configure Wireshark to
+read. Wireshark is then able to decrypt the traffic, which is required to see some packate details (for example, the
+OCSP Staple response details)
+```aidl
+-javaagent:$ROOTDIR/jSSLKeyLog.jar=jssl-key.log 
+```
+
+Similarly, it is also possible to trace the GRPC/Protobuf traffic by configuring the
+Protobuf add-in Wireshark to use the Daml protobuf files. You may need to download Daml
+repo and Google base grpc protobuf files for this to work successfully.
```

Comment:
Worth mentioning that this is probably a bad idea for performance, security, … and should only be used for debugging?